solr/core/src/java/org/apache/solr/search/SegmentMap.java

```diff
@@ -0,0 +1,207 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.search;
+
+import java.util.AbstractMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+
+/**
+ * A lightweight metadata representation of the segment layout of a particular view ({@link
+ * SolrIndexSearcher} over an index. This may be used, e.g., to capture the context of stale cache
+ * entries, where portions of cached values may still be valid over newer views of the index.
+ *
+ * <p>The metadata contained here can be used to leverage stale cache entries in the efficient
+ * reconstruction of new cache entries by "re-mapping" still-valid segment ranges of the old cache
+ * entry onto the new cache entry (thereby avoiding the need to actually <i>compute</i> the new
+ * cached value for that segment range.
+ */
+public class SegmentMap {
+
+  private static final AtomicLong IDS = new AtomicLong();
+
+  /**
+   * Generates a unique (classloader-scoped) String id to be associated with generated segment maps.
+   * This enables scoping of objects and data structures that are associated with a given segment
+   * map.
+   *
+   * <p>This may be too cute in its construction of ids, but since we have sequentially-assigned
+   * ids, it's preferable that ids vary in their <i>lower</i> indexes (to avoid having to compare
+   * many character indexes). {@link Character#MAX_RADIX} and reversing the string accomplishes
+   * this.
+   */
+  // TODO: could we just use
+  //  SolrIndexSearcher.getTopReaderContext().reader().getReaderCacheHelper().getKey() and use
+  //  simple {@link Object} for scoping instead? ... and perhaps fallback to String when such is
+  //  not available?
+  private static String mintId() {
+    long raw = IDS.getAndIncrement();
+    String inOrder = Long.toUnsignedString(raw, Character.MAX_RADIX);
+    return new StringBuilder(inOrder).reverse().toString();
+  }
+
+  public static SegmentMap generateSegmentMap(SolrIndexSearcher searcher) {
+    final List<LeafReaderContext> leafContexts = searcher.getLeafContexts();
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    Map.Entry<IndexReader.CacheKey, Segment>[] segs = new Map.Entry[leafContexts.size()];
+    int i = 0;
+    for (LeafReaderContext ctx : leafContexts) {
+      LeafReader r = ctx.reader();
+      IndexReader.CacheKey coreKey = r.getCoreCacheHelper().getKey();
```

Comment:
I believe LeafReaderContext refers to per lucene segment? If that's the case should this be a segment key? Im a bit confused cause each coreKey here seems to make to a segment?